### PR TITLE
Implement name inference for mm, addmm, addmm_

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25181 [WIP][testing] Turn on BUILD_NAMEDTENSOR=1
* #25178 Include the correct header for make_unique in named tensor headers
* #25177 Implement name inference for torch.matmul
* #25123 Implement name inference for torch.bmm
* #24474 Implement name inference for torch.dot
* #24471 Implement name inference for addmv, addmv_, mv
* #24469 Implement name inference for expand
* **#24306 Implement name inference for mm, addmm, addmm_**
* #24921 Fix binary op name inference between unnamed and named tensors.

Featuring:
- a new way of writing name inference tests. At some point I'll migrate
the older tests over.

Test Plan:
- new tests [namedtensor ci]

gh-metadata: pytorch pytorch 24306 gh/zou3519/104/head

Differential Revision: [D16915801](https://our.internmc.facebook.com/intern/diff/D16915801)

### CODE REVIEW

**Review by nairbv — APPROVED** [review]

File: test/test_namedtensor.py [hunk 1]
```diff
@@ -23,10 +24,58 @@ def flatten(lst):
...
 class TestNamedTensor(TestCase):
     def test_trivial(self):
         pass
 
+    def _test_name_inference(self, op, args=(), expected_names=(), device='cpu',
+                             maybe_raises_regex=None):
+        casted_args = [arg.to(device) if isinstance(arg, torch.Tensor) else arg
+                       for arg in args]
+        if maybe_raises_regex is not None:
+            with self.assertRaisesRegex(RuntimeError, maybe_raises_regex):
```
Comment (nairbv) [inline test/test_namedtensor.py]: i don't see this used... should we test some cases that fail (e.g. mis-matched names in a mm op should fail right?)

Comment (zou3519) [reply test/test_namedtensor.py]: Right now, the name inference rule for mm doesn't fail if names are mismatched, so that won't fail.

Comment (nairbv) [reply test/test_namedtensor.py]: is it supposed to fail?

Comment (zou3519) [reply test/test_namedtensor.py]: The decision right now is for it to not fail. Tensor[M, N] @ Tensor[O, P] returns Tensor[M, P] and it doesn't matter that N / O are "mismatched".

Comment (nairbv) [reply test/test_namedtensor.py]: ah now I understand, thanks

**Pull request closed without merging**